Remove `v` from kubecf's version string, and consume correctly

viccuad opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Remove `v` char from the semver string, to be fully compliant with semver, both in bazel and shell based build systems.

This changes the resulting artifact name from the build; adjust regexps and shell expansion in pipeline
to consume them. 

If merged, this would surely trigger changes downstream for consuming the tars and version string.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Another implementation could be to not change the artifact name, and only the yaml value in `Chart.yaml`. This seems hackish, and given that the call to building with helm is `helm package  --version "${VERSION}" …`, seems foolish.

Closes https://github.com/cloudfoundry-incubator/kubecf/issues/927.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- An important detail to include is the version of the used cf-operator -->

Deployed pipeline from this branch at https://concourse.suse.dev/teams/main/pipelines/kubecf-viccuad?group=viccuad-semver-full.

Tested also by rebasing this branch against release-2.3, setting `version := 2.3.0`, and running up until smokes-diego. 

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code has security implications.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

viccuad commented:
Rebased master for the linting fixes.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -419,11 +421,11 @@ jobs:
   - put: s3.kubecf-ci
     params:
-      file: output/kubecf-v*.tgz
+      file: output/kubecf-[0-9]*.tgz
```
Shouldn't this contain at least `.`? I think we'd get git commit IDs and things like that, too (for non-release builds).

viccuad replied:
I don't think so. Note that this is not a regexp, but shellscript expansion: first a number, then whatever.

This change is motivated by being able to select the file `kubecf-0.0.0-f541d0f.tgz` but not select `kubecf-bundle-0.0.0-f541d0f.tgz`. Since we don't have the `v0.0.0` anymore, nor lookahead in shellscript expansion, we need something like this.

viccuad commented:
Should not be merged before https://github.com/cloudfoundry-incubator/kubecf/issues/1095 is solved, or we break CI for release-2.2, release-2.3 branches.

jandubois commented:
> Should not be merged before #1095 is solved, or we break CI for release-2.2, release-2.3 branches.

Turned the PR into draft mode to prevent accidental merging.

viccuad commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -133,7 +133,8 @@ resources:
     region_name: {{ $config.s3_bucket_region }}
-    regexp: kubecf-v(.*).tgz
+    # adapted from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+    regexp: kubecf-(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?.tgz
```
While it works as it is, it is true that there's no need to have so many capture groups.

The s3 resource uses the first capture group to obtain the full version, and uses that version as the s3 versioning of the file (which we don't use).

I have changed the regexps to have:
- only 1 regexp group: correct for s3
- full regexp for semver: allows us to implicitly check that our versions adhere to semvers

viccuad commented:
Thanks for the review Mark!

Rebased on top of master. keeps being blocked by the fix in https://github.com/cloudfoundry-incubator/kubecf/pull/1108.

> We should probably also fix the [packaging script](https://github.com/cloudfoundry-incubator/kubecf/blob/74773f1ce0ecae59f2b8bcd4db2380284c2072a9/rules/helm/package.tmpl.rb#L65) where it generates a version for non-release builds.

Fixed with https://github.com/cloudfoundry-incubator/kubecf/pull/1083/commits/f541d0f48c6f8e08097553f403f4d59791556048.

Resolved asked changes.

viccuad commented:
Unblocked after merging #1111, rebased on master to be able to deploy a working pipeline to update the github status here.

viccuad commented:
Can be tested by approving the PR, pinning this PR number in https://concourse.suse.dev/teams/main/pipelines/kubecf-viccuad/resources/kubecf-pr, and then unpausing the `lint-pr` job here: https://concourse.suse.dev/teams/main/pipelines/kubecf-viccuad?group=pr. That will update the github statuses here.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -127,7 +127,9 @@ resources:
-    regexp: kubecf-v(.*).tgz
+    # adapted from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+    # with 1 capturing group for the whole semver, as the s3 resource expects
+    regexp: kubecf-[0-9]+\.[0-9]+\.[0-9]+(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?(?:\+[0-9A-Za-z-]+)?.tgz
```
As far as I can tell, there's no longer a capturing group at all (all groups are non-capturing, `(?:`).
```suggestion
    regexp: kubecf-([0-9]+\.[0-9]+\.[0-9]+(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?(?:\+[0-9A-Za-z-]+)?).tgz
```

viccuad replied:
Totally true, I managed to mangle my last interactive rebase. Pushed a new commit.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -136,7 +138,9 @@ resources:
-    regexp: kubecf-bundle-v(.*).tgz
+    # adapted from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+    # with 1 capturing group for the whole semver, as the s3 resource expects
+    regexp: kubecf-bundle-[0-9]+\.[0-9]+\.[0-9]+(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?(?:\+[0-9A-Za-z-]+)?.tgz
```
Same here.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -145,7 +149,7 @@ resources:
     secret_access_key: ((aws-secret-key))
     region_name: {{ $config.s3_final_bucket_region }}
-    regexp: kubecf-v(.*).tgz
+    regexp: kubecf-(.*).tgz
```
Does this match `kubecf-bundle-*`?

viccuad replied:
It does not, solved in last commit.

viccuad commented:
Thanks for the review, they were indeed all problems.
Solved and flown again in https://concourse.suse.dev/teams/main/pipelines/kubecf-viccuad. This time seems it works correctly.

thardeck commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -2035,13 +2039,13 @@ jobs:
     {{- end }}
     params:
-      file: output/kubecf-bundle-v*.tgz
+      file: output/kubecf-bundle-*.tgz
```
Is there a reason why the number glob is missing here?

viccuad replied:
not really, it's just not needed. It will shell-expand to the number of the semver anyways, we don't have another file like kubecf-bundle-foo.tgz.
